### PR TITLE
Issue 3825: Pravega security vulnerability reporting policy

### LINKED ISSUE #3825: A page in Pravega Website about reporting security vulnerabilities
**Problem description**
We need a page in Pravega Website that explains how to report security vulnerabilities. (security@pravega.io)

**Problem location**
Pravega Website

**Suggestions for an improvement**

### PR DESCRIPTION
**Pull request by ravibeta**
**Change log description**  
This change adds security.md page as described by [https://github.com/github/opensource.guide](https://github.com/github/opensource.guide/issues/981) for the purpose of including a security page for Pravega.

**Purpose of the change**  
Fixes #3825 

**What the code does**  
The DellEMC products follow a security configuration guide that lists the security vulnerabilities reporting method. This change includes the proper references, links, and descriptions to enable researchers to file security vulnerabilities against Pravega.

**How to verify it**

### CODE REVIEW

File: SECURITY.md [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+-->
+<!-- BEGIN DellEMC SECURITY.MD V0.0.1 BLOCK -->
+
+## Security
+
+DellEMC takes the security of our software products and services seriously, which includes all source code repositories managed through our GitHub organizations, which include [Pravega](https://github.com/pravega/pravega).
```
Comment (ravisharda) [inline SECURITY.md:14]: This and other statements in this document have ingredients of a proprietary corporate policy and references a company. It should be agnostic to a particular company, and shouldn't be referencing Dell EMC in my view.

File: SECURITY.md [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+If you believe you have found a security vulnerability in any DellEMC-owned repository that meets [DellEMC's definition of a security vulnerability](https://www.dell.com/support/contents/en-us/article/product-support/self-support-knowledgebase/security-antivirus/alerts-vulnerabilities/dell-vulnerability-response-policy), please report it to us as described below.
+
+## Reporting Security Issues
+
+**Please do not report security vulnerabilities through public GitHub issues.**
+
+Instead, please report them to the DellEMC Product Security Incident Response Team (Dell PSIRT) via email(mailto:secure@dell.com).
```
Comment (ravisharda) [inline SECURITY.md:22]: Pravega-specific email of the form - `...@pravega.io`

File: SECURITY.md [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+
+If possible, encrypt your message with our PGP key; please download it from the [Dell Vulnerability Response Policy page](https://i.dell.com/sites/csdocuments/eSupport_Docs/en/dell_psirt_pub_key.txt).
```
Comment (ravisharda) [inline SECURITY.md:24]: This might not be applicable as-is if you use a `@pravega.io` email ID.

**Comment by derekm** [comment]
> I thought we had a `security@pravega.io` address for that

We definitely have that email address. The recipients are:
>      "s.owens@dell.com",
>      "flavio.junqueira@emc.com",
>      "arvind.kandhare@emc.com",
>      "Ravi.Sharda@emc.com",

Behind the scenes, we may need policies around sharing these reports with SDP and other vendors that rely on Pravega.

**Comment by ravisharda** [comment]
@ravibeta could you please remove the reference to Dell EMC products from the PR description as well. 

Also, "How to Verify" section points to the file in your fork. Please remove that reference.

File: SECURITY.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+Pravega takes the security of our software products and services seriously, which includes all source code repositories managed through our GitHub organizations.
+
+If you believe you have found a security vulnerability in any Pravega owned repository, please report it to us as described below.
+
+## Reporting Security Issues
+
+**Please do not report security vulnerabilities through public GitHub issues.**
+
+Instead, please report them to the Pravega Security Team via email(mailto:security@pravega.io).
```
Comment (ravisharda) [inline SECURITY.md:22]: Fix formatting of `email(mailto:security@pravega.io).`

File: SECURITY.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+You should receive a response within 24 hours. If for some reason you do not, please follow up via email to ensure we received your original message.
```
Comment (ravisharda) [inline SECURITY.md:24]: I am not sure about "receive a response within 24 hours". I don't think we should promise that. These two lines could be removed.

File: SECURITY.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+Please include the requested information listed below (as much as you can provide) to help us better understand the nature and scope of the possible issue:
+
+  * Product name and version containing the vulnerability
```
Comment (ravisharda) [inline SECURITY.md:28]: Why product name?

File: SECURITY.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+  * Type of issue (e.g. buffer overflow, SQL injection, cross-site scripting, etc.)
+  * Full paths of source file(s) related to the manifestation of the issue
+  * The location of the affected source code (tag/branch/commit or direct URL)
```
Comment (ravisharda) [inline SECURITY.md:31]: Is this different from the previous line?

File: SECURITY.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+This information will help us triage your report more quickly.
+
+
+## Preferred Languages
+
+We prefer all communications to be in English.
+
+## Policy
+
+Pravega follows the principle of [Coordinated Vulnerability Disclosure](https://www.cisa.gov/coordinated-vulnerability-disclosure-process).
```
Comment (ravisharda) [inline SECURITY.md:46]: Have you got a confirmation about this? cc @fpj / @derekm .

File: SECURITY.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+<!-- END PRAVEGA SECURITY.MD BLOCK -->
+
+## Released Versions
+The following versions of Pravega open source are released. Please visit the [Releases](https://github.com/pravega/pravega/releases) page for tags and changelists. 
+
+| Version | Released          |
```
Comment (ravisharda) [inline SECURITY.md:53]: It'd be hard to maintain the versions here. Just referencing the releases page should be enough.

File: SECURITY.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+| ------- | ------------------ |
+| 0.7.x   | :white_check_mark: |
+| 0.6.x   | :white_check_mark: |
+| < 0.6   | :x:                |
+
+Pravega is 100% open source and community-driven. All components are available under [Apache 2 License](https://www.apache.org/licenses/LICENSE-2.0.html) on GitHub.
```
Comment (andreipaduroiu) [inline SECURITY.md:59]: This statement is redundant. It's clearly stated on our home page.

File: SECURITY.md (hunk 8 above)
Comment (derekm) [reply SECURITY.md:46]: @ravisharda -- I think that's a holdover from when @ravibeta was reusing Dell's policy. Declaring alignment with the CVD process may make sense.

As for Core Infrastructure Initiative's Best Practices guidelines at the "passing" level, our policy must at least meet these criteria:
* The project's initial response time for any vulnerability MUST be less than or equal to 14 days.
* There MUST be no unpatched vulnerabilities of medium or higher severity that have been publicly known for more than 60 days. [Severity determined by Common Vulnerability Scoring System (CVSS).]
* Projects SHOULD fix all critical vulnerabilities rapidly after they are reported.
* It is SUGGESTED that at least one of the static analysis tools used for the static_analysis criterion include rules or approaches to look for common vulnerabilities in the analyzed language or environment.
* All medium and higher severity exploitable vulnerabilities discovered with static code analysis MUST be fixed in a timely way after they are confirmed.
* All medium and higher severity exploitable vulnerabilities discovered with dynamic code analysis MUST be fixed in a timely way after they are confirmed.

File: SECURITY.md (hunk 6 above)
Comment (derekm) [reply SECURITY.md]: e.g., Schema Registry

File: SECURITY.md (hunk 5 above)
Comment (derekm) [reply SECURITY.md]: CII Best Practices stipulates 14 days to respond: https://github.com/pravega/pravega/pull/5052#discussion_r472357416

File: SECURITY.md [hunk 11]
```diff
@@ -19,20 +19,19 @@ If you believe you have found a security vulnerability in any Pravega owned repo
...
-  * Product name and version containing the vulnerability
-  * Type of issue (e.g. buffer overflow, SQL injection, cross-site scripting, etc.)
-  * Full paths of source file(s) related to the manifestation of the issue
-  * The location of the affected source code (tag/branch/commit or direct URL)
+  * Product version containing the vulnerability
+  * Type of security exploit (e.g. buffer overflow, SQL injection, cross-site scripting, etc.)
+  * The location of the vulnerability in the source code (tag/branch/commit or direct URL)
   * Any special configuration required to reproduce the issue
-  * Step-by-step instructions to reproduce the issue
+  * Step-by-step instruction to reproduce the issue
```
[the remark this answers is not in the document]
Comment (ravibeta) [reply SECURITY.md:32]: Yes Derek.

File: SECURITY.md [hunk 12]
```diff
@@ -19,20 +19,19 @@ If you believe you have found a security vulnerability in any Pravega owned repo
 
 **Please do not report security vulnerabilities through public GitHub issues.**
 
-Instead, please report them to the Pravega Security Team via email(mailto:security@pravega.io).
+Instead, please report them to the Pravega Security Team via `email(mailto:security@pravega.io)`.
```
[the remark this answers is not in the document]
Comment (ravibeta) [reply SECURITY.md:22]: Yes Derek.

File: SECURITY.md [hunk 13]
```diff
@@ -19,20 +19,19 @@ If you believe you have found a security vulnerability in any Pravega owned repo
...
-You should receive a response within 24 hours. If for some reason you do not, please follow up via email to ensure we received your original message.
+You should receive a response within 14 days. If for some reason you do not, please follow up via email to ensure we received your original message.
 
 Please include the requested information listed below (as much as you can provide) to help us better understand the nature and scope of the possible issue:
 
-  * Product name and version containing the vulnerability
-  * Type of issue (e.g. buffer overflow, SQL injection, cross-site scripting, etc.)
-  * Full paths of source file(s) related to the manifestation of the issue
-  * The location of the affected source code (tag/branch/commit or direct URL)
+  * Product version containing the vulnerability
```
[the remark this answers is not in the document]
Comment (ravibeta) [reply SECURITY.md:28]: Yes Derek.

**Comment by andreipaduroiu** [comment]
Closing this for now. We may reopen if we decide to continue this work.

**Pull request closed without merging**